### PR TITLE
[Backend modularization] `metabase.server.core` and `metabase.request.core` API namespaces; update Kondo

### PR DESCRIPTION
**Pull request by camsaul**
Split the stuff related to Ring request/the current request (used all over the app) into a new `metabase.request` module.  Most of the stuff related to the current request lives here now (such as current limit, offset, and tools like `with-current-user` and `as-admin`). Now other namespaces can use `request/` instead of a mix of random server middleware namespaces, and this simplifies the dependency graph between things a lot.

`metabase.server.core` is the new web server API namespace; most things don't need to use this anymore.

I want to move stuff like `*current-user-id*` into the `metabase.request` module as well but that will be a pretty big refactor so I'll leave that as a follow-on to this project.

This PR required untangling a lot of knots and circular dependencies but I think I'm pretty happy with the progress made here. Still more work to do after this lands, tho.

### CODE REVIEW

File: src/metabase/request/core.clj [hunk 1]
```diff
@@ -0,0 +1,53 @@
+(ns metabase.request.core
+  "API namespace for stuff related to Ring (HTTP) requests, such as info associated with the current request."
+  (:require
+   [metabase.request.cookies]
+   [metabase.request.current]
+   [metabase.request.session]
+   [metabase.request.util]
+   [potemkin :as p]))
```
Comment (bshepherdson) [inline src/metabase/request/core.clj:8]: Please use `metabase.util.namespaces/import-fns` just for uniformity. It expands to `p/import-vars` in CLJ, so you still get `doc` and go-to-definition just like Potemkin.

File: src/metabase/server/core.clj [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  public-exceptions]
+  ;; TODO -- I think all of this stuff probably belongs in [[metabase.request.*]]
+ [metabase.server.protocols
+  Respond])
+
+(defn handler
+  "Get the top-level Ring handler."
+  []
+  ;; dynamically resolved for now because [[metabase.server.handler]] depends on [[metabase.api.routes]] and thus would
+  ;; make a big old circular deps MESS...
```
Comment (bshepherdson) [inline src/metabase/server/core.clj:37]: `TODO` to clean this up?

File: src/metabase/server/instance.clj [hunk 3]
```diff
@@ -0,0 +1,130 @@
+(ns metabase.server.instance
```
Comment (bshepherdson) [inline src/metabase/server/instance.clj:1]: I'm not seeing a deleted file this code came from - did you miss a `git rm`?

This seems like a lot of pretty core machinery not to have existed already.

File: src/metabase/server/middleware/misc.clj [hunk 4]
```diff
@@ -79,13 +79,11 @@
...
 (defn bind-request
   "Ring middleware that binds `*request*` for the duration of this Ring request."
   [handler]
   (fn [request respond raise]
-    (binding [*request* request]
-      (handler request respond raise))))
+    (request/do-with-current-request
+     request
+     (^:once fn* []
+       (handler request respond raise)))))
```
Comment (bshepherdson) [inline src/metabase/server/middleware/misc.clj:89]: This seems like it deserves a basic macro to wrap up the body in the `(^:once fn* [] ...)`. It seems like the usual `do-with-foo` function + `with-foo` macro pattern is in play, but skipping the macro part.

**Review by bshepherdson — APPROVED** [review]

File: .clj-kondo/config.edn [hunk 5]
```diff
@@ -442,7 +452,7 @@
     metabase.troubleshooting       :any
     metabase.types                 #{metabase.util}
     metabase.upload                :any
-    metabase.util                  :any
+    metabase.util                  :any ; TODO -- this should probably be #{}
```
[the remark this answers is not in the document]
Comment (camsaul) [reply .clj-kondo/config.edn:454]: No, this is the part of the config that says what modules a module is allowed to depend on. `:any` says that we're not enforcing any rules there (yet). Ideally I think `metabase.util` would be a foundational "brick" that wouldn't depend on anything else. That's a problem for another day tho

File: src/metabase/server/instance.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/server/instance.clj:1]: Oops, good catch. This is renamed from `metabase.server`. Fixed

File: src/metabase/server/middleware/misc.clj (hunk 4 above)
Comment (camsaul) [reply src/metabase/server/middleware/misc.clj:89]: Yeah I probably should, I just didn't want to deal with the Kondo config nonsense honestly. Laziness is not really a good excuse tho. I will fix

File: src/metabase/server/core.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/server/core.clj]: done.

File: src/metabase/server/middleware/misc.clj (hunk 4 above)
Comment (camsaul) [reply src/metabase/server/middleware/misc.clj]: fixed

File: src/metabase/request/core.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/request/core.clj:8]: I'll open a GH issue to fix all of the usages

**Pull request merged**